icechunk/src/ops/gc.rs

```diff
@@ -0,0 +1,296 @@
+use std::{collections::HashSet, future::ready, iter};
+
+use chrono::{DateTime, Utc};
+use futures::{stream, Stream, StreamExt, TryStreamExt};
+use tokio::pin;
+
+use crate::{
+    format::{ChunkId, ManifestId, SnapshotId},
+    refs::{list_refs, RefError},
+    repository::ChunkPayload,
+    storage::ListInfo,
+    Storage, StorageError,
+};
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Action {
+    Keep,
+    DeleteIfCreatedBefore(DateTime<Utc>),
+}
+
+#[derive(Debug)]
+pub struct GCConfig {
+    extra_roots: HashSet<SnapshotId>,
+    dangling_chunks: Action,
+    dangling_manifests: Action,
+    dangling_attributes: Action,
+    dangling_transaction_logs: Action,
+    dangling_snapshots: Action,
+}
+
+impl GCConfig {
+    pub fn new(
+        extra_roots: HashSet<SnapshotId>,
+        dangling_chunks: Action,
+        dangling_manifests: Action,
+        dangling_attributes: Action,
+        dangling_transaction_logs: Action,
+        dangling_snapshots: Action,
+    ) -> Self {
+        GCConfig {
+            extra_roots,
+            dangling_chunks,
+            dangling_manifests,
+            dangling_attributes,
+            dangling_transaction_logs,
+            dangling_snapshots,
+        }
+    }
+    pub fn clean_all(
+        chunks_age: DateTime<Utc>,
+        metadata_age: DateTime<Utc>,
```

Comment:
Curious about the rationale for having two different ages here.

Reply:
I thought, giving that metadata files are far less numerous than chunks, users would be willing to preserve them longer. If we someday support some form of reflog (or even without) it could be useful to have them around for forensics or understanding. It wouldn't allow to recover data, but you could at least recover "structure". For example, a reasonable thing to do could be to GC chunks older than 1 month and metadata older than 1 year.